src/Microsoft.Health.Fhir.SqlServer/Features/Storage/Registry/SqlServerSearchParameterStatusDataStore.cs

```diff
@@ -0,0 +1,116 @@
+            _updateSearchParamsTvpGenerator = updateSearchParamsTvpGenerator;
+            _filebasedSearchParameterStatusDataStore = filebasedRegistry.Invoke();
+            _schemaInformation = schemaInformation;
+        }
+
+        // TODO: Make cancellation token an input.
+        public async Task<IReadOnlyCollection<ResourceSearchParameterStatus>> GetSearchParameterStatuses()
+        {
+            // If the search parameter table in SQL does not yet contain status columns
+            if (_schemaInformation.Current < SchemaVersionConstants.SearchParameterStatusSchemaVersion)
+            {
+                // Get status information from file.
+                return await _filebasedSearchParameterStatusDataStore.GetSearchParameterStatuses();
+            }
+
+            using (IScoped<SqlConnectionWrapperFactory> scopedSqlConnectionWrapperFactory = _scopedSqlConnectionWrapperFactory())
+            using (SqlConnectionWrapper sqlConnectionWrapper = await scopedSqlConnectionWrapperFactory.Value.ObtainSqlConnectionWrapperAsync(CancellationToken.None, true))
+            using (SqlCommandWrapper sqlCommandWrapper = sqlConnectionWrapper.CreateSqlCommand())
+            {
+                VLatest.GetSearchParamStatuses.PopulateCommand(sqlCommandWrapper);
+
+                var parameterStatuses = new List<ResourceSearchParameterStatus>();
+
+                using (SqlDataReader sqlDataReader = await sqlCommandWrapper.ExecuteReaderAsync(CommandBehavior.SequentialAccess, CancellationToken.None))
+                {
+                    while (await sqlDataReader.ReadAsync())
+                    {
+                        (string uri, string stringStatus, DateTimeOffset? lastUpdated, bool? isPartiallySupported) = sqlDataReader.ReadRow(
+                            VLatest.SearchParam.Uri,
+                            VLatest.SearchParam.Status,
+                            VLatest.SearchParam.LastUpdated,
+                            VLatest.SearchParam.IsPartiallySupported);
+
+                        var status = Enum.Parse<SearchParameterStatus>(stringStatus, true);
+
+                        var resourceSearchParameterStatus = new ResourceSearchParameterStatus()
+                        {
+                            Uri = new Uri(uri),
+                            Status = status,
+                            IsPartiallySupported = (bool)isPartiallySupported,
+                            LastUpdated = (DateTimeOffset)lastUpdated,
+                        };
+
+                        parameterStatuses.Add(resourceSearchParameterStatus);
+                    }
+                }
+
+                return parameterStatuses;
+            }
+        }
+
+        // TODO: Make cancellation token an input.
+        public async Task UpsertStatuses(IEnumerable<ResourceSearchParameterStatus> statuses)
+        {
+            EnsureArg.IsNotNull(statuses, nameof(statuses));
```

Comment:
>statuses [](start = 49, length = 8)

I would check is this collection empty or not, before doing all this SQL with it.

Reply:
Good idea, added in 12919af.